File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.py
n,m =map(int, input().split())
r,c,d = map(int, input().split())
mp = [list(map(int, input().split())) for _ in range(n)]

# 북, 동, 남, 서
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

res = 0
curx = r
cury = c
curd = d
while(True):
    
    if(mp[curx][cury] == 0):
        mp[curx][cury] = 2
        res += 1

    need_clean = False    
    for _ in range(4):
        curd = (curd + 3) % 4
        nx = curx + dx[curd]
        ny = cury + dy[curd]

        if(nx >= 0 and ny >= 0 and nx < n and ny < m and mp[nx][ny] == 0):
            need_clean = True
            curx = nx
            cury = ny
            break
    
    if need_clean:
        continue

    back_dir = (curd + 2) % 4
    bx = curx + dx[back_dir]
    by = cury + dy[back_dir]

    if (bx >= 0 and by >= 0 and bx < n and by < m and mp[bx][by] == 1):
        break
    
    curx = bx
    cury = by

print(res)

File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.cc
#include <iostream>
#include <vector>
using namespace std;
vector<vector<int>>map(50, vector<int>(50));
int dx[4] = { 1,-1,0,0 };
int dy[4] = { 0,0,1,-1 };
int n, m;

int solve(int x, int y, int d) {
	int res = 0;
	int curd = d;
	int curx = x;
	int cury = y;
	while (true) {
		if (map[curx][cury] == 0) {
			map[curx][cury] = 2;
			res++;
		}
		bool need_clean = false;
		for (int i = 0; i < 4; i++) {
			int nx;
			int ny;
			if (curd == 0) {
				nx = curx;
				ny = cury - 1;
			}
			else if (curd == 1) {
				nx = curx - 1;
				ny = cury;
			}
			else if (curd == 2) {
				nx = curx;
				ny = cury + 1;
			}
			else {
				nx = curx + 1;
				ny = cury;
			}
			curd = (curd + 3) % 4;
			if (nx >= 0 && ny >= 0 && nx < n && ny < m && map[nx][ny] == 0) {
				curx = nx;
				cury = ny;
				need_clean = true;
				break;
			}
		}
		if (need_clean) continue;

		int nx;
		int ny;
		if (curd == 0) {
			nx = curx + 1;
			ny = cury;
		}
		else if (curd == 1) {
			nx = curx;
			ny = cury - 1;
		}
		else if (curd == 2) {
			nx = curx - 1;
			ny = cury;
		}
		else {
			nx = curx;
			ny = cury + 1;
		}
		if (nx >= 0 && ny >= 0 && nx < n && ny < m && map[nx][ny] == 1 ) {
			return res;
		}
		curx = nx;
		cury = ny;
	}
}
int main() {
	cin >> n >> m;
	int r, c, d; cin >> r >> c >> d;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> map[i][j];
		}
	}
	int ans = solve(r, c, d);
	cout << ans;
}